docs/sparse_boundary_communication.md

```diff
@@ -0,0 +1,131 @@
+
+In practice, we denote each channel by a unique key `std::tuple<int, int, std::string, int>{send_gid, receive_gid, var_name, geometric_element_index_of_sender}`, so that the `Mesh` can contain a map from these keys to communication channels. Then, at each remesh, sending blocks and blocks that are receiving from blocks on a different rank can create new communication channels and register them in this map. *Implementation detail:* To build these keys, we currently rely on the `MeshBlock::std::unique_ptr<BoundaryValues> pbval` object to get information about the neighboring blocks and build the channels and keys. `BoundaryValues` has its own communication methods defined, but none of these are used for the sparse communication. We really only rely on the information stored in `BoundaryBase` (which contains general information about all of the neighboring blocks on the mesh), which `BoundaryValues` inherits from. Eventually, I think `pbval` should be turned into a `BoundaryBase` object and `BoundaryValues` should be removed from the code base.
+
+### MPI Communication IDs
+For MPI commnunication, we need to uniquely identify MPI messages between an ordered pair of ranks. In an easy world, we would do all communication on `MPI_COMM_WORLD` and just create a unique tag from the elements of the unique key tuple by doing something like `tag = num_blocks_total * nvars * 27 * send_gid + nvars * 27 * receive_gid + 27 * var_idx + geometric_element_index_of_sender`, but this creates sparsely populated tag numbers and quickly blows through the maximum tag limit of 32767 defined by the MPI standard. In practice, we create an MPI communicator associated with each variable (see `Mesh::SetupMPIComms()` and `Mesh::GetMPIComm(std::string)`) and for each pair of ranks fill a `std::map<UnorderedPair({send_gid, send_geom_elem}, {recv_gid, recv_geom_elem}), int>` contained in the `Mesh` with zeros. Since `std::map` is always sorted and we define comparators for `UnorderedPair`, we end up with the same ordering of the map on both ranks of the pair. We can then iterate through the map and assign a unique, densely populated tag to each pair in the map, which does not blow through the MPI tag limit. The same tags can obviously be re-used in each communicator. 
```

Comment:
The way it works now, we would have to have more than 32767 shared geometric elements (i.e. faces, edges, or vertices) between the blocks on one rank and another distinct rank to blow through the number of MPI tags (since there are no tag numbers that go unused). If the actual MPI upper bound is larger than that, we can certainly go to a larger number of shared geometric elements between two ranks. When I messed around with it `MPI_TAB_UB` seemed to give strange results some time so I just gave up on it. The actual values of `gid` and `lid` for the blocks don't play a direct role in determining the tag number (so there is no more limit like 1024). Rather, they are just used as parts of a map key. After the map is filled with all needed keys, it can be sorted, and then assigned tags sequentially.

Reply:
Ah, I see. Way better design than before!